compiler/compiler/src/lib.rs

```diff
@@ -102,20 +103,45 @@ impl<'a> Compiler<'a> {
             program_string,
         )?;
         // Write the AST snapshot post parsing.
-        ast.to_json_file_without_keys(self.output_directory, "initial_ast.json", &["span"])?;
+        ast.to_json_file_without_keys(self.output_directory.clone(), "initial_ast.json", &["span"])?;
+
+        self.ast = ast;
+
+        Ok(())
+    }
+
+    // Loads the given input file if it exists.
+    fn parse_input(&self, input_file_path: PathBuf) -> Result<Option<ParsedInputFile>> {
```

Comment:
Is there a reason why `parse_input` returns an `Option` and `parse_program` returns `()`?

Reply:
I could change this if you want, but parse_program attaches the Ast to the Compiler, and parse_input returns it? The `parse_input` is called from `compiler_stages`  and, therefore, the parent `compile` function rather than public.

Reply:
either `parse_input` should assign to `self.input` and return `Result<()>`
or `parse_program` should return an `Ast`